pkg/gatewayserver/io/io.go

```diff
@@ -661,3 +681,44 @@ func (c *Connection) notifyStatsChanged() {
 	default:
 	}
 }
+
+func uplinkMessageFromProto(pb *ttnpb.UplinkMessage) *uplinkMessage {
+	up := &uplinkMessage{
+		payload:   pb.GetRawPayload(),
+		frequency: pb.GetSettings().Frequency,
+	}
+	for _, md := range pb.GetRxMetadata() {
+		up.antennaMask = up.antennaMask | (1 << md.GetAntennaIndex())
+	}
+	return up
+}
+
+func isRepeatedUplink(this *uplinkMessage, that *uplinkMessage) bool {
+	switch {
+	case this == nil || that == nil:
+		return false
+	case this.frequency != that.frequency:
+		return false
+	case this.antennaMask != that.antennaMask:
+		return false
+	case !bytes.Equal(this.payload, that.payload):
+		return false
+	}
+	return true
+}
+
+// Interval between emitting consecutive gs.up.repeat events for the same gateway connection.
+const consecutiveRepeatUpEventsInterval = time.Minute
+
+func (c *Connection) checkRepeatedUplink(up *uplinkMessage) (repeated, shouldEmitEvent bool) {
```

Comment:
nit; let the caller (`HandleUp()`) not bother with the data structure (`uplinkMessage`) that is being used internally.

Reply:
This is also not just a check (which implies no state is changed), whereas state (`lastUp`) is actually being changed. This is important for our future selves; this method isn't goroutine safe.